hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/SlidingWindowCalcSplitLogicalRule.java

```diff
@@ -71,21 +70,32 @@ public void onMatch(RelOptRuleCall call) {
         final CalcLogicalRel calc = call.rel(0);
         final SlidingWindow sw = call.rel(1);
 
+        final boolean[] mustNotExecute = {false};
+
         RexProgram program = calc.getProgram();
         RexVisitorImpl<Void> visitor = new RexVisitorImpl<Void>(true) {
             @Override
             public Void visitInputRef(RexInputRef ref) {
                 int index = ref.getIndex();
                 if (index == sw.windowStartIndex() || index == sw.windowEndIndex()) {
-                    // TODO[sasha]: convert into ShouldNotExecuteRel (after #20996 merge)
-                    throw QueryException.error("Can't apply filter criteria to window bounds");
+                    mustNotExecute[0] = true;
                 }
                 return super.visitInputRef(ref);
             }
         };
 
         program.expandLocalRef(program.getCondition()).accept(visitor);
 
+        if (mustNotExecute[0]) {
+            call.transformTo(new MustNotExecuteLogicalRel(
```

Comment:
Shouldn't we use `SHOULD-not-execute` here? We need to avoid the "not enough rules" error, so we ensure that this rule converts to something, but if there's another rule that can execute the plan, it should be applied. But I'm not 100% sure...

Reply:
Notice, that `MustNotExecuteLogicalRel` has zero cost, and it is supposed to be the best plan always.